### ISSUE TITLE
BayesSearchCV weird performance

### ISSUE THREAD

**Issue by fingoldo** [scikit-optimize/scikit-optimize#703]
Hello, this topic is about comparison of performance of sklearn's RandomizedSearchCV, GridSearchCV and the skopt's BayesSearchCV.
We have discussed this issue with Tim more than half year ago, at that time skopt had problems installing on Windows, now it hasn't and I had a chance to run the test again.

Results are still far from what we could expect from a theoretically superior method, can anyone shed some light on why it's taking so much time to use BayesSearchCV compared to RandomizedSearchCV, under ANY base_estimator?

_Note: I had to reduce number of trials for BayesSearchCV to 40 instead of original 216 to get a feasible run time._
```

import numpy as np
import pandas as pd
from time import time

from sklearn.datasets import load_digits
from scipy.stats import randint as sp_randint
from sklearn.ensemble import RandomForestClassifier

from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import GridSearchCV,RandomizedSearchCV

from skopt import BayesSearchCV
from skopt.space import Real, Categorical, Integer

#inits
pd.set_option('max_colwidth',400)

# Utility function to report best scores+timings
def report_perf(optimizer, X, y, title):
    from IPython.display import display
    start = time()
    optimizer.fit(X, y)
    d=pd.DataFrame(optimizer.cv_results_)
    print((title+" took %.2f seconds,  candidates checked: %d, best CV score: %.3f "+u"\u00B1"+" %.3f")
              % (time() - start, len(optimizer.cv_results_['params']),d[d.rank_test_score==1]['mean_test_score'].values[0],d[d.rank_test_score==1]['std_test_score'].values[0]))    
    display(d[['rank_test_score','mean_test_score','std_test_score','params']]
            .sort_values(by='mean_test_score',ascending =False).head(5))

# get some data
digits = load_digits()
X, y = digits.data, digits.target
#define partitioning
skf = StratifiedKFold(n_splits=5,random_state=22)
# build a base classifier
clf = RandomForestClassifier(n_estimators=20,random_state=32)

random_search = RandomizedSearchCV(clf, param_distributions={"max_depth": [3, None],
              "max_features": sp_randint(1, 11),
              "min_samples_split": sp_randint(2, 11),
              "min_samples_leaf": sp_randint(1, 11),
              "bootstrap": [True, False],
              "criterion": ["gini", "entropy"]},n_iter=216,n_jobs=-1,cv=skf,
                                   return_train_score=False)
report_perf(random_search, X, y,'RandomizedSearchCV')

grid_search = GridSearchCV(clf, param_grid={"max_depth": [3, None],
              "max_features": [1, 3, 10],
              "min_samples_split": [2, 3, 10],
              "min_samples_leaf": [1, 3, 10],
              "bootstrap": [True, False],
              "criterion": ["gini", "entropy"]},n_jobs=-1,cv=skf,return_train_score=False)
report_perf(grid_search, X, y,'GridSearchCV')

for baseEstimator in ['GP','RF','ET']:
    opt = BayesSearchCV(clf,
        {"max_depth": Categorical([3, 100000]),
                  "max_features": (1, 10),
                  "min_samples_split": (2, 10),
                  "min_samples_leaf": (1, 10),
                  "bootstrap": [True, False],
                  "criterion": ["gini", "entropy"]},
        n_iter=40,n_jobs=-1,cv=skf,return_train_score=False,optimizer_kwargs={'base_estimator': baseEstimator})
    report_perf(opt, X, y,'BayesSearchCV_'+baseEstimator)
```

My results:
```

RandomizedSearchCV took 14.51 seconds,  candidates checked: 216, best CV score: 0.938 ± 0.013
GridSearchCV took 14.52 seconds,  candidates checked: 216, best CV score: 0.932 ± 0.029

BayesSearchCV_GP took 146.50 seconds,  candidates checked: 40, best CV score: 0.913 ± 0.033
BayesSearchCV_RF took 146.48 seconds,  candidates checked: 40, best CV score: 0.807 ± 0.026
BayesSearchCV_ET took 147.24 seconds,  candidates checked: 40, best CV score: 0.808 ± 0.025


```

Do not be worried about absolute accuracy not being high enough, as we are feeding the model with raw pixels without any preprocessing or dimensionality reduction.
But what is drawing attention, is huge runtime of BayesSearchCV.
Why it it taking so long to fit BayesSearchCV? In my understanding, should not be training set for it pretty small? As inputs to this metaestimator we (in this case) should have 6 columns and up to 40 training examples, and, roughly speaking, should perform up to 40 fits of the metaestimator (in fact, even less, as it would be wise to use some batches instead of re-fitting it after every new trial).
But running of 40 fits on 6 columns of maximum 40 rows SHOULD NOT take 2 minutes! I'd expect maximum 1 second, not more.
So what is happening?

More on this, if we make n_iter =40 for RandomizedSearchCV to be able to compare its efficiency directly, we can see it's outperforming BayesSearchCV!!!
```

RandomizedSearchCV took 10.40 seconds,  candidates checked: 40, best CV score: 0.923 ± 0.034
RandomizedSearchCV took 10.44 seconds,  candidates checked: 40, best CV score: 0.932 ± 0.022
RandomizedSearchCV took 10.55 seconds,  candidates checked: 40, best CV score: 0.932 ± 0.020
```

So not only is BayesSearchCV much slower, it's also arriving at worse solutions than random search. What is going on??

Regards,
Anatoly

**Comment by iaroslav-ai** [comment]
Hi @fingoldo ,
here are some ideas:
- `scikit-optimize` is focused on optimizing model parameters, where a single fitting of the model takes considerable amount of time, e.g. hours or more. This is done using [Bayesian Optimization](https://papers.nips.cc/paper/4522-practical-bayesian-optimization-of-machine-learning-algorithms.pdf) (BO), as this class of algorithms has a property that it can find optimal hyperparameters of a model in relatively small number of trials of different parameter configurations. This means, that a smaller amount of training time is necessary (more on that below). However, such algorithm is not very suited for problems where model fitting takes very small amount of time, e.g. microseconds. The reason is that Bayesian Optimization requires fitting of a "surrogate" function, which models how cross - validation score changes w.r.t. different parameters. This is done every time a new parameter values are tried to see in what cross - validation they result. Fitting of this surrogate function is costly and results exactly in the overhead that you are seeing. However, this surrogate function is what leads to the favorable properties of these algorithms. For long running fitting, such costs are negligible. 
- I would re-run the experiment a few times to see how results vary from run to run - there is some randomness in how BO works. Also in some situations random search might be better than BO, simply because of [no free lunch theorem](http://georgemaciunas.com/wp-content/uploads/2012/07/Wolpert_NLFoptimization-1.pdf). But it might be also that with larger number BO performs equally or even better. But if BO is consistently outperformed by random search, this should become a great test for our algorithms! If you have time for some experimentation with different number of iterations for example, it would be great if you check this. 
- It is interesting that you are seeing this warning. It means that at some iteration an algorithm came up with a parameter configuration to evaluate that has been already evaluated. Could you post a complete log of BayesSearchCV? That might give an idea of whether there could be any issue. In particular, I would look how often such warning is raised.

**Comment by fingoldo** [comment]
Thank you for reply Iaroslav, so it's supposed to take so much time? incredible.
Anyway something must be wrong as it's not converging to solution better than found by random search, given the same amount of iterations and not taking execution time into account.
Results I have posted are very repeatable, same estimates I had half year ago.
And when I run the script again I get about the same results.

Do you have any example at hand where Skopt shows its superiority over random search, which one could run more or less easily, like the one I have provided?

So you think that particular example I have stumbled upon is exactly the case according to No free lunch theorem where random search is better than BO? ;-)

Regarding complete log of BayesSearchCV, let me know please how to extract it and i will provide it.

**Comment by iaroslav-ai** [comment]
Hello Anatoly.
I have modified the evaluation part in your example a bit and used also `None` as a value for `max_depth` in BayesSearchCV:
```python
import numpy as np
import pandas as pd
from time import time

from sklearn.datasets import load_digits
from scipy.stats import randint as sp_randint
from sklearn.ensemble import RandomForestClassifier

from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import GridSearchCV,RandomizedSearchCV

from skopt import BayesSearchCV
from skopt.space import Real, Categorical, Integer

# Utility function to report best scores+timings
def report_perf(optimizer, X, y, title):
    optimizer.fit(X, y)
    print(title, "best CV score:", optimizer.best_score_)

# get some data
digits = load_digits()
X, y = digits.data, digits.target
#define partitioning
skf = StratifiedKFold(n_splits=5,random_state=22)
# build a base classifier
clf = RandomForestClassifier(n_estimators=20,random_state=32)

total_iter = 40

for baseEstimator in ['GP','RF','ET']:
    opt = BayesSearchCV(clf,
        {"max_depth": Categorical([3, None]),
                  "max_features": (1, 10),
                  "min_samples_split": (2, 10),
                  "min_samples_leaf": (1, 10),
                  "bootstrap": [True, False],
                  "criterion": ["gini", "entropy"]},
        n_iter=total_iter,n_jobs=-1,cv=skf,return_train_score=False,optimizer_kwargs={'base_estimator': baseEstimator})
    report_perf(opt, X, y,'BayesSearchCV_'+baseEstimator)

random_search = RandomizedSearchCV(clf, param_distributions={"max_depth": [3, None],
              "max_features": sp_randint(1, 11),
              "min_samples_split": sp_randint(2, 11),
              "min_samples_leaf": sp_randint(1, 11),
              "bootstrap": [True, False],
              "criterion": ["gini", "entropy"]},n_iter=total_iter,n_jobs=-1,cv=skf,
                                   return_train_score=False)
report_perf(random_search, X, y,'RandomizedSearchCV')

grid_search = GridSearchCV(clf, param_grid={"max_depth": [3, None],
              "max_features": [1, 3, 10],
              "min_samples_split": [2, 3, 10],
              "min_samples_leaf": [1, 3, 10],
              "bootstrap": [True, False],
              "criterion": ["gini", "entropy"]},n_jobs=-1,cv=skf,return_train_score=False)
report_perf(grid_search, X, y,'GridSearchCV')
```
I run this with `python3 -W ignore skopt_test.py` to suppress the warnings, and I am consistently getting that BayesSearchCV is beating the other methods; This is what I get in console:
```bash
BayesSearchCV_GP best CV score: 0.9376739009460211
BayesSearchCV_RF best CV score: 0.9309961046188091
BayesSearchCV_ET best CV score: 0.9254312743461325
RandomizedSearchCV best CV score: 0.9298831385642737
GridSearchCV best CV score: 0.9315525876460768
```
I use the version of `skopt` on master branch. 

Will look into this more when I am back. Can you reproduce these results?

**Comment by iaroslav-ai** [comment]
Hi again :) Here are some thoughts also:

Regarding the logs for BayesSearchCV - I got it. The internal optimization algorithm suggests same points to try once the algorithm is converging, and hence it throws the warning. In this case it seems it can be safely ignored.

We have benchmarked skopt against other methods on a set of ~50+ optimization tasks, see [here](https://github.com/iaroslav-ai/scikit-optimize-benchmarks). These results suggest that indeed `gp_minimize` is quite performant, compared to random search. Some benchmarks can be found in [here](https://github.com/scikit-optimize/scikit-optimize/blob/master/skopt/benchmarks.py).

I will look tomorrow / on weekend into why I am getting different results when using `best_score_` then what you got.

**Comment by fingoldo** [comment]
Thank you Slava so much for finding time to clear this question.
Yes when I run your example BO is showing better accuracies )

2018-07-26 16:02:46.856298 Starting fit of BayesSearchCV_GP
2018-07-26 16:05:52.414450 BayesSearchCV_GP best CV score: 0.9376739009460211 {'bootstrap': False, 'criterion': 'gini', 'max_depth': None, 'max_features': 6, 'min_samples_leaf': 1, 'min_samples_split': 2}
2018-07-26 16:05:52.421456 Starting fit of BayesSearchCV_RF
2018-07-26 16:09:00.612143 BayesSearchCV_RF best CV score: 0.9343350027824151 {'bootstrap': False, 'criterion': 'gini', 'max_depth': None, 'max_features': 8, 'min_samples_leaf': 1, 'min_samples_split': 4}
2018-07-26 16:09:00.619149 Starting fit of BayesSearchCV_ET
2018-07-26 16:12:14.007575 BayesSearchCV_ET best CV score: 0.9376739009460211 {'bootstrap': False, 'criterion': 'gini', 'max_depth': None, 'max_features': 8, 'min_samples_leaf': 1, 'min_samples_split': 2}
2018-07-26 16:12:14.016083 Starting fit of RandomizedSearchCV
2018-07-26 16:12:18.421895 RandomizedSearchCV best CV score: 0.9265442404006677 {'bootstrap': False, 'criterion': 'entropy', 'max_depth': None, 'max_features': 6, 'min_samples_leaf': 3, 'min_samples_split': 9}
2018-07-26 16:12:18.422396 Starting fit of GridSearchCV
2018-07-26 16:12:25.925234 GridSearchCV best CV score: 0.9315525876460768 {'bootstrap': False, 'criterion': 'entropy', 'max_depth': None, 'max_features': 10, 'min_samples_leaf': 1, 'min_samples_split': 3}.

Hope I will have time soon to test it on real problem where one pass over data takes comparatively longer to justify BO run time. Thank you again!

**Comment by fingoldo** [comment]
However, I would be really interested to compare performance of SkOpt with some simple heuristics, like: 1) sample less from a region if it's taking larger times to evaluate target function in that region. Is this taken into account somehow by current implementation of BayesSearchCV?
2) use kind of multiarmed bandit heuristic, where at start we sample randomly and with time we sample mostly from the best regions, only occasionally sampling from distant regions to explore new opportunities (exploration-exploitation dilemma)
3) as i am particularly interested in machine learning applications, decrease probability of sampling values of a parameter which has resulted in estimator's fit failure, due to unsupported combination of parameters. Sklearn's methods currently dumbly repeat sampling in such cases over and over again, wasting cycles. Don't know how BayesSearchCV behaves.

Also, I still don't know why it's taking BO so long to decide on next step. As far as i understand it, it's trying to interpolate probabilities distribution over some grid in search space? Maybe number of points over which it's doing that can be adjusted by some extra knob? So that users can set it to lower value and it can compete with RandomSearchCV not only in solution quality but also in run time?

**Comment by iaroslav-ai** [comment]
Hi, here are some ideas:
1) We actually have a functionality that takes into account time, see [here](https://github.com/scikit-optimize/scikit-optimize/blob/000b9d848c4afd5cc12852c3df0b1b76476d00b4/skopt/optimizer/optimizer.py#L87) - you are then required to use a special type of acquisition function for the optimizer used in `BayesSearchCV`. To provide such configuration, you can use `optimizer_kwargs` of the `BayesSearchCV`.
2) It is interesting that you mentioned that; I actually have an implementation of a special random exploration function [here](https://github.com/optimization-js/optimization-js/blob/64c7ff7c39471dde9cce09e8bf9735770829d305/python_prototypes/ga.py#L161), which performs surprisingly well, and it behaves approximately how you suggest - big changes are introduced only rarely. It is not a part of scikit - optimize though. 
3) If you provide some `error_score <= 0`, I think BayesSearchCV should learn after some time which configurations are infeasible (lead to minimal score) and hence to not use them. But to compare to others, I would set up some kind of a small scale experiment to see if this is true. 

One of the costly operations in BO is calculation of a kernel function between all pairs of observations, which already gives O(N^2) calculations for N observations. There are other operations, which increase the cost of evaluation even more. There is a [variant of BO algorithm](http://proceedings.mlr.press/v37/snoek15.pdf), which is more scalable. But is still might take more time computationally than random search.

The interesting thing is that if you can evaluate a lot of hyperparameter configurations in short amount of time, maybe you do not need BO in the first place. A typical situation is that solution produced by BayesSearchCV in N iterations will be also achieved by RandomSearchCV in 2*N or 4*N or 8*N iterations. If evaluation of hyperparameter configuration is very cheap relative to BO step, then you just spend extra second for more random evaluations, and get similar result.  

An interesting "in - between" solution would be some advanced genetic algorithm. Here to support this: in the recent black box competition, in [single objective track](https://bbcomp.ini.rub.de/results/BBComp2018-1OBJ/summary.html) a combination of genetic and local search works best; but in [expensive single objective](https://bbcomp.ini.rub.de/results/BBComp2018-1OBJ-expensive/summary.html) gaussian processes base method is working best.
